src/Microsoft.Extensions.Configuration.AzureAppConfiguration/ConfigurationClientManager.cs

```diff
@@ -0,0 +1,118 @@
+
+using Azure.Core;
+using Azure.Data.AppConfiguration;
+using Microsoft.Extensions.Configuration.AzureAppConfiguration.Constants;
+using Microsoft.Extensions.Configuration.AzureAppConfiguration.Extensions;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.Extensions.Configuration.AzureAppConfiguration
+{
+    /// <summary>
+    /// A configuration client provider which maintains state of configuration clients and provides set of clients to use when requested.
+    /// </summary>
+    /// <remarks>
+    /// This class is not thread-safe. Since config provider does not allow multiple network requests at the same time,
+    /// there won't be multiple threads calling this client at the same time.
+    /// </remarks>
+    internal class ConfigurationClientManager : IConfigurationClientManager
+    {
+        private readonly IList<ConfigurationClientStatus> _clients;
+        private readonly EndpointUriComparer _endpointComparer = new EndpointUriComparer();
+
+        public ConfigurationClientManager(string connectionString, ConfigurationClientOptions clientOptions)
+        {
+            if (string.IsNullOrEmpty(connectionString))
+            {
+                throw new ArgumentNullException(nameof(connectionString));
+            }
+
+            var endpoint = new Uri(ConnectionStringParser.Parse(connectionString, ConnectionStringParser.EndpointSection));
+            var configurationClientStatus = new ConfigurationClientStatus(endpoint, new ConfigurationClient(connectionString, clientOptions));
+            _clients = new List<ConfigurationClientStatus> { configurationClientStatus };
+        }
+
+        public ConfigurationClientManager(IEnumerable<Uri> endpoints, TokenCredential credential, ConfigurationClientOptions clientOptions)
+        {
+            if (endpoints == null || !endpoints.Any())
+            {
+                throw new ArgumentNullException(nameof(endpoints));
+            }
+
+            _clients = endpoints.Select(endpoint => new ConfigurationClientStatus(endpoint, new ConfigurationClient(endpoint, credential, clientOptions))).ToList();
+        }
+
+        /// <summary>
+        /// Internal constructor; Only used for unit testing.
+        /// </summary>
+        /// <param name="clients"></param>
+        internal ConfigurationClientManager(IList<ConfigurationClientStatus> clients)
+        {
+            _clients = clients;
+        }
+
+        public IEnumerable<ConfigurationClient> GetAvailableClients()
+        {
+            IEnumerable<ConfigurationClient> clients = Enumerable.Empty<ConfigurationClient>();
+
+            clients = _clients.Where(client => client.BackoffEndTime <= DateTimeOffset.UtcNow).Select(c => c.Client);
+
+            // If all clients are in the back-off state, try all clients anyways.
+            if (!clients.Any())
+            {
+                clients = _clients.Select(c => c.Client);
+            }
+
+            return clients;
```

Comment:
Should we retrun `clients.ToList()`? Otherwise, if someone calls `GetAvailableClients` repeatedly, we will reevaluate the LINQ every time.

Reply:
Sorry, I gave a bad example. If someone uses the returned clients repeatedly, the LINQ will be evaluated repeatedly every time the returned clients are referenced. This is not a problem for your current code because the returned clients are only referenced once (so Jimmy is right that it's not necessary for the current code). However, our intention for this function is to evaluate the LINQ once per GetAvailableClients call. We don't expect someone to get the returned clients and use it so the LINQ result can change dynamically. So calling ToList() matches better the design of this function.

Reply:
There is a question of whether not repeated enumerations of an enumerable should return consistent results. I don't think the answer needs to be yes, but if we do want that behavior then there are a couple options

* Use ToList
* Declare DateTimeOffset.UtcNow outside of `Where`

I usually opt to avoid short-lived collections when possible, so I'd go for the second option.

By the way, the current code does have a problem because `clients.Any` may be true when invoked on line 65, but may be false when caller enumerates, resulting in possible empty enumeration.

Reply:
What is the problem with short-lived collections?
It's good to have a definitive list of clients, right?

Reply:
Updated it to have `.ToList()`.